### PR TITLE
De-duplicate indentations in JSX Texts

### LINKED ISSUE #20766: JSX formatting does not de-indent closing tag with text body
From https://github.com/Microsoft/vscode/issues/40454#issuecomment-352566521

**TypeScript Version:**  2.7.0-dev.20171214

**Code**

```ts
(
    <div>
        text
        </div>
)
```

Run format

**Expected behavior:**

```js
(
    <div>
        text
    </div>
)
```

**Actual behavior:**
No change. Content remains:

```js
(
    <div>
        text
        </div>
)
```

The `</div>` is de-indented if the body is a tag instead of text:

```js
(
    <div>
        <span />
        </div>
)
```

### PR DESCRIPTION
**Pull request by orta**
Fixes #20766 - Some of the interesting notes are in https://github.com/orta/TypeScript/pull/1

### CODE REVIEW

File: src/services/formatting/formatting.ts [hunk 1]
```diff
@@ -736,10 +741,22 @@ namespace ts.formatting {
                 const childIndentation = computeIndentation(child, childStartLine, childIndentationAmount, node, parentDynamicIndentation, effectiveParentStartLine);
 
                 processNode(child, childContextNode, childStartLine, undecoratedChildStartLine, childIndentation.indentation, childIndentation.delta);
-
                 if (child.kind === SyntaxKind.JsxText) {
                     const range: TextRange = { pos: child.getStart(), end: child.getEnd() };
-                    indentMultilineCommentOrJsxText(range, childIndentation.indentation, /*firstLineIsIndented*/ true, /*indentFinalLine*/ false);
+                    if (range.pos !== range.end) { // don't indent zero-width jsx text
+                        const siblings = parent.getChildren(sourceFile);
+                        const currentNode = find(siblings, arg => arg.pos === child.pos);
```
Comment (andrewbranch) [inline src/services/formatting/formatting.ts]:
```suggestion
                        const currentIndex = findIndex(siblings, arg => arg.pos === child.pos);
```

Comment (orta) [reply src/services/formatting/formatting.ts]: Thanks, pushing change up now

**Comment by orta** [comment]
@typescript-bot cherry-pick this to branchname release-3.8

**Comment by orta** [comment]
@typescript-bot cherry-pick this to release-3.8

**Comment by weswigham** [comment]
@typescript-bot cherry-pick this to release-3.8

**Comment by DanielRosenwasser** [comment]
@typescript-bot cherry-pick this to release-3.8 and LKG

**Comment by orta** [comment]
Thanks, was just coming to do that

**Comment by DanielRosenwasser** [comment]
Woah, hold up, this thing doesn't make PRs anymore @weswigham?

**Comment by weswigham** [comment]
Uhhh, that's news to me, especially since I had it open a PR (https://github.com/microsoft/TypeScript/pull/36644) just the other day after it was updated.

**Comment by weswigham** [comment]
I think I'm being lied to and that @orta manually cherry-picked it, since the cherry-pick script in this branch is out of date :P (y'know, since the last build for this branch is a [failure](https://typescript.visualstudio.com/TypeScript/_build/results?buildId=63413&view=logs&j=fd490c07-0b22-5182-fac9-6d67fe1e939b&t=7cc137c2-61f2-5d65-b5f8-0e03d1af3170))

**Pull request merged**